contracts/BitGuildMarketplace.sol

```diff
@@ -0,0 +1,334 @@
+
+        emit LogListingExtended(seller, _contract, _tokenId, listings[key].createdAt, listings[key].expiry);
+
+        return true;
+    }
+
+    // @dev Withdraw item from marketplace back to seller
+    // @param _contract whitelisted contract
+    // @param _tokenId  tokenId
+    function withdrawItem(address _contract, uint _tokenId) public onlyWhitelisted(_contract) {
+        bytes32 key = _getHashKey(_contract, _tokenId);
+        address seller = listings[key].seller;
+
+        require(seller == msg.sender, "Only seller can withdraw listing.");
+
+        if (!allowWithdrawBeforeExpiry) {
+            require(listings[key].expiry <= now, "Withdraw only available after expired.");
+        }
+
+        // Transfer item back to the seller
+        ERC721 gameToken = ERC721(_contract);
+        gameToken.safeTransferFrom(this, seller, _tokenId);
+
+        emit LogItemWithdrawn(seller, _contract, _tokenId, now);
+
+        // remove listing
+        _delist(key);
+    }
+
+    // ===========================================
+    // Purchase Item
+    // ===========================================
+    // @dev Buy item with ETH. Take ETH from buyer, transfer token, transfer payment minus fee to seller
+    // @param _contract  Token contract
+    // @param _tokenId   Token Id
+    function buyWithETH(address _contract, uint _tokenId) public onlyWhitelisted(_contract) payable {
+        bytes32 key = _getHashKey(_contract, _tokenId);
+        uint price = listings[key].price;
+        address seller = listings[key].seller;
+        Currencies currency = listings[key].currency;
+
+        require(currency == Currencies.ETH, "Listing not in ETH.");
+        require(msg.value > 0 && msg.value == price, "Invalid price.");
+        require(listings[key].expiry > now, "Item expired.");
+
+        ERC721 gameToken = ERC721(_contract);
+        require(gameToken.ownerOf(_tokenId) == address(this), "Item is not available.");
+
+        uint fee;
+        (,fee) = getFee(price, msg.sender, seller, _contract); // getFee returns percentFee and fee, we only need fee
+
+        // Transfer item token to buyer
+        gameToken.safeTransferFrom(this, msg.sender, _tokenId);
+        // Transfer Balance - fee to Seller
+        require(seller.send(price - fee) == true, "Transfer to seller failed.");
+
+        // delist item
+        _delist(key);
+
+        // Emit event
+        emit LogItemSold(msg.sender, seller, _contract, _tokenId, price, currency, now);
+    }
+
+    // Buy with PLAT requires calling BitGuildToken contract, this is the callback
+    // call to approve already verified the token ownership, no checks required
+    // @param _buyer     buyer
+    // @param _value     PLAT amount (big number)
+    // @param _PLAT      BitGuild token address
+    // @param _extraData address _gameContract, uint _tokenId
+    function receiveApproval(address _buyer, uint _value, BitGuildToken _PLAT, bytes _extraData) public {
```

Comment:
This is VERY uncool. A lot of dupe code! Abstract away. The difference here is like in 2-3 lines of code

Reply:
Abstracted out.